### ISSUE TITLE
End to End ML Project Procedure

### ISSUE THREAD

**Issue by phymo** [phymooc/learn-AI#1]
# AI

## 1 分类

监督学习，非监督学习，半监督学习

分类，线性回归，强化学习（奖励）

自然语言处理，图像语音识别

批量学习，增量学习

## 2. 步骤

- 洞察
- 分析 数据清理 选择模型
- 实现 调参
- 发布

## 3. 工具

- 分析数据： numpy pandas
- 实现：
    - tensorflow pytorch 底层
    - keras 高层 常用  深度学习
    - scikit-learn 集成常用算法 model 和 dataset
    - spark: 基于Hadoop 常用与大数据 和 生产
- 发布： kubeflow airflow ai_suit(naver)

## 4. 数据分析

- 指标： 均方根误差（标准差 RMSE)， 平均绝对误差，L1范数 L2 范数
- 检验假设： 一些方法（head, info, describe, hist(直方图））

## 5. 创建测试集

- 抽样： 20%等比例分层抽样

## 6. 寻找相关性

- corr: Correlation： 1. 试验属性间的相关性，2. 试验属性组合与结果的相关性

## 7. 数据清理 Normalization

- 对于缺失的特征值： 删掉，或者插值（0 或者 平均值）
- 文本处理：1. 转数字（有问题）， 2. 独热编码 （二进制 oneHotEncode)
- 特征缩放： 1. 最大最小缩放 （0-1归一化）， 2. 标准化： standardScaler

## 8. keras 框架 and Scikit-Learn

- 估算器
- 转换器
- 预测器

- 几个方法： fit（训练）， predict，transform

## 10. 选择和训练模型

- 算法： 线性回归，决策树，随机森林 等
- 评估： 欠拟合，过拟合， 交叉验证（不同等级）
- 参数微调（fine tune): 1. 网格搜索：grid search （分层试参数）， 2. 随机搜索， 3. 集成方法

## 11. 启动和部署

部署到 aisuit(naver) 或者 kubeflow

通过restful API + web app 对外提供服务

## 12. 监控：

1. Grafana.， Prometheus
